### DISCUSSION TITLE
Filesystem scan error when passing in custom policies

### DISCUSSION THREAD

**Discussion by AnaisUrlichs** [aquasecurity/trivy#4668] [Bugs]
### Description

I try to run the following command:
```
trivy config --policy ./custom-policies/ignore/basic-two.rego --namespaces user ./kubernetes-manifests
```

The trivy config scan works on the Kubernetes manifests without the `--policy` flag. 

### Desired Behavior

I have used the Rego before. Usually, Trivy would perform a scan independent on whether the Rego was correct or not -- this in turn, allowed me to change the Rego.

### Actual Behavior

Once I pass in the Rego policy, I get the following output:
```
2023-06-19T14:22:47.892+0100	INFO	Misconfiguration scanning is enabled
2023-06-19T14:22:47.895+0100	FATAL	filesystem scan error: scan error: unable to initialize a scanner: unable to initialize a filesystem scanner: analyzer group error: post-analyzer init error: dockerfile scanner init error: mapfs file copy error: file does not exist
```

### Reproduction Steps

```bash
1. rego file

package trivy

import data.lib.trivy

default ignore = false

ignore {
	input.CweIDs[_] == "CWE-787"
}
```
2. Kubernetes YAML
https://github.com/Cloud-Native-Security/trivy-demo/tree/main/manifests

3. Command
```
trivy config --policy ./custom-policies/ignore/basic-two.rego --namespaces user ./kubernetes-manifests
```

Normal fs scanning works

### Target

Kubernetes

### Scanner

Misconfiguration

### Output Format

Table

### Mode

Standalone

### Debug Output

```bash
2023-06-19T14:23:55.135+0100	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-06-19T14:23:55.153+0100	DEBUG	cache dir:  /Users/anaisurlichs/Library/Caches/trivy
2023-06-19T14:23:55.153+0100	INFO	Misconfiguration scanning is enabled
2023-06-19T14:23:55.154+0100	DEBUG	Policies successfully loaded from disk
2023-06-19T14:23:55.158+0100	FATAL	filesystem scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:430
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:268
  - unable to initialize a scanner:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:677
  - unable to initialize a filesystem scanner:
    github.com/aquasecurity/trivy/pkg/commands/artifact.filesystemStandaloneScanner
        github.com/aquasecurity/trivy/pkg/commands/artifact/scanner.go:59
  - analyzer group error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/local.NewArtifact
        github.com/aquasecurity/trivy/pkg/fanal/artifact/local/fs.go:55
  - post-analyzer init error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer.NewAnalyzerGroup
        github.com/aquasecurity/trivy/pkg/fanal/analyzer/analyzer.go:363
  - terraform scanner init error:
    github.com/aquasecurity/trivy/pkg/fanal/analyzer/config.NewAnalyzer
        github.com/aquasecurity/trivy/pkg/fanal/analyzer/config/config.go:34
  - mapfs file copy error:
    github.com/aquasecurity/trivy/pkg/misconf.createPolicyFS
        github.com/aquasecurity/trivy/pkg/misconf/scanner.go:297
  - file does not exist
```

### Operating System

macOS Ventura

### Version

```bash
❯ trivy version                                                                                                 
Version: 0.42.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-06-19 06:08:54.473589677 +0000 UTC
  NextUpdate: 2023-06-19 12:08:54.473589076 +0000 UTC
  DownloadedAt: 2023-06-19 10:35:25.09426 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2023-03-22 00:58:27.964412713 +0000 UTC
  NextUpdate: 2023-03-25 00:58:27.964412013 +0000 UTC
  DownloadedAt: 2023-03-22 10:15:37.261165 +0000 UTC
Policy Bundle:
  Digest: sha256:030345bff4845bae1c89450da85681bc25c0c40d2144a0d9db5254674d155298
  DownloadedAt: 2023-06-19 11:24:14.243634 +0000 UTC
```

### Checklist

- [ ] Run `trivy --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by simar7** [comment]
It seems to be fine for me:

```shell
trivy config --policy ./custom-policies --namespaces user ~/repos/trivy-demo/kubernetes-manifests | head -n
15
2023-06-19T16:43:14.656-0600	INFO	Misconfiguration scanning is enabled
2023-06-19T16:43:15.751-0600	INFO	Detected config files: 2

deployment.yaml (kubernetes)
============================
Tests: 141 (SUCCESSES: 132, FAILURES: 9, EXCEPTIONS: 0)
Failures: 9 (UNKNOWN: 0, LOW: 8, MEDIUM: 1, HIGH: 0, CRITICAL: 0)

LOW: Container 'react-application' of Deployment 'react-application' should add 'ALL' to 'securityContext.capabilities.drop'
════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
The container should drop all default capabilities and add only those that are needed for its execution.

See https://avd.aquasec.com/misconfig/ksv003
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 deployment.yaml:16-26
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  16 ┌       - name: react-application

<snip>
```

```shell
tree custom-policies
custom-policies
└── mypolicy.rego

0 directories, 1 file
➜  4668 cat custom-policies/mypolicy.rego
package trivy

import data.lib.trivy

default ignore = false

ignore {
	input.CweIDs[_] == "CWE-787"
}
```

Please note `--policy` accepts a path for the top level directory that contains all the custom policies. Maybe we can document this better? Thoughts?

**Reply by knqyf263** [reply]
Or rename it to `--policy-diir`? We took the flag name and behavior from Conftest, but it may be unclear.

**Reply by AnaisUrlichs** [reply]
That does not make much sense to me -- so you are saying I HAVE to provide a specific directory? 
If that is the case, will update the docs -- I just would have assumed I can pass in the specific rego file :D 
I seem to have always used a dir before

<img width="1039" alt="Screenshot 2023-06-20 at 10 41 06" src="https://github.com/aquasecurity/trivy/assets/33576047/ef2ffbf0-4b74-4360-9d7c-0cb2cbf029b8">

**Reply by simar7** [reply]
As Teppei mentioned the behaviour and flag names are from conftest. We can always change the docs or the flag names if we feel they are a better fit for us.

For now I suggest we update the docs first. Renaming an existing flag is backwards breaking unless we introduce a new flag (an alias).

**Reply by knqyf263** [reply]
@simar7 Do you think we can add support for files? We probably need to detect whether the specified path is a file or a directory.

**Reply by simar7** [reply]
Yeah we could do that. Issue to track: https://github.com/aquasecurity/trivy/issues/4689

**Reply by itaysk** [reply]
1. i don't think we should keep referring to conftest we are not compatible with it since long time ago
2. if the flag behaves differently in fs target as you said then i'd call it a bug no?
3. if we are renaming the flag let's consider the best name more carefully (https://github.com/aquasecurity/trivy/discussions/3620)

**Reply by knqyf263** [reply]
>1. i don't think we should keep referring to conftest we are not compatible with it since long time ago

We're just talking about the reason why it behaves so originally. We tried to mimic the behavior of Conftest. Nothing more, nothing less. We're not trying to keep the compatibility anymore.

>2. if the flag behaves differently in fs target as you said then i'd call it a bug no?

It works as expected. Files are not supported, so this is a feature request.

>3. if we are renaming the flag let's consider the best name more carefully (https://github.com/aquasecurity/trivy/discussions/3620)

We won't rename the flag, but update the behavior.

**Reply by itaysk** [reply]
> It works as expected. Files are not supported, so this is a feature request.

@AnaisUrlichs wrote "Normal fs scanning works" with the file as argument.

**Reply by AnaisUrlichs** [reply]
I think you two are talking about two different things (correct me if I'm wrong or misunderstood)

I can run 
> trivy fs < directory I want to scan >

This is the normal fs scanning that I referred to

However, what does not work is 
> trivy fs --policy <directory to policy/specific policy file> < directory I want to scan >

Instead, trivy policy needs to have the directory passed in, it cannot accept a specific file e.g.
> trivy fs --policy < directory to policy > < directory I want to scan >

**Reply by AnaisUrlichs** [reply]
https://github.com/aquasecurity/trivy/pull/4716

**Reply by knqyf263** [reply]
>Normal fs scanning works

As @AnaisUrlichs said, it means it works without passing policies.
